### PR TITLE
Add parquet support

### PR DESCRIPTION
**Pull request by meln1k**
# Description

Added support for parquet export

# To-Dos

- [ ] Add test coverage for new or updated functionality
- [ ] Lint and test with `tox`
- [ ] Document new or updated functionality (someengineering/resoto.com#XXXX)

# Code of Conduct

By submitting this pull request, I agree to follow the [code of conduct](https://resoto.com/code-of-conduct).

### CODE REVIEW

File: cloud2sql/parquet.py [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+}
+
+
+carz = [
+    Property("cloud", "string"),
+    Property("account", "string"),
+    Property("region", "string"),
+    Property("zone", "string"),
+]
+carz_access = {name: ["ancestors", name, "reported", "id"] for name in ["cloud", "account", "region", "zone"]}
```
Comment (aquamatthias) [inline cloud2sql/parquet.py]: Maybe we should move this to a separate model module - it does not belong to SQL or Parquet, but the source data we have.

File: cloud2sql/parquet.py [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+        self.schemas: Dict[str, pa.Schema] = {}
+
+    def table_name(self, kind: str) -> str:
+        replaced = kind.replace(".", "_")
+        return replaced
+
+    def link_table_name(self, from_kind: str, to_kind: str) -> str:
+        # 63 characters to be consistent with SQL model
+        replaced = f"link_{self.table_name(from_kind)[0:25]}_{self.table_name(to_kind)[0:25]}"
+        return replaced
```
Comment (aquamatthias) [inline cloud2sql/parquet.py]: If we want to be consistent - maybe we should move this logic also to a separate module.

File: cloud2sql/parquet.py [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+            table.append(values)
+            self.table_content[table_name] = table
+            return True
+        return False
+
+    def insert_node(self, node: JsObject) -> None:
+        if node.get("type") == "node" and "id" in node and "reported" in node:
+            reported: Json = node.get("reported", {})
+            for name, value in reported.items():
+                if isinstance(value, dict) or isinstance(value, list) or isinstance(value, set):
```
Comment (aquamatthias) [inline cloud2sql/parquet.py]:
```suggestion
                if isinstance(value, (dict, list, set)):
```

File: cloud2sql/parquet.py [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+                    reported[name] = json.dumps(value)  # convert to json in case of a nested object
+
+            reported["_id"] = node["id"]
+            reported["cloud"] = value_in_path(node, carz_access["cloud"])
+            reported["account"] = value_in_path(node, carz_access["account"])
+            reported["region"] = value_in_path(node, carz_access["region"])
+            reported["zone"] = value_in_path(node, carz_access["zone"])
+            kind = reported.pop("kind")
+            self.kind_by_id[node["id"]] = kind
```
Comment (aquamatthias) [inline cloud2sql/parquet.py]: I also see the same code in the SQL case - we should DRY it up.

File: cloud2sql/parquet.py [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+        self.model = model
+        self.kind_by_id: Dict[str, str] = {}
+        self.table_content: Dict[str, List[Dict[str, Any]]] = {}
+
+    def insert_value(self, table_name: str, values: Any) -> bool:
+        if self.model.schemas.get(table_name):
+            table = self.table_content.get(table_name, [])
+            table.append(values)
+            self.table_content[table_name] = table
+            return True
```
Comment (aquamatthias) [inline cloud2sql/parquet.py]: when is the data written to disk?

File: cloud2sql/collect_plugins.py [hunk 6]
```diff
@@ -56,7 +59,59 @@ def configure(path_to_config: Optional[str]) -> Json:
...
+    for node in collector.graph.nodes:
+        node._graph = collector.graph
+        exported = node_to_dict(node)
+        exported["type"] = "node"
+        exported["ancestors"] = {
+            "cloud": {"reported": {"id": node.cloud().name}},
+            "account": {"reported": {"id": node.account().name}},
+            "region": {"reported": {"id": node.region().name}},
+            "zone": {"reported": {"id": node.zone().name}},
+        }
```
Comment (aquamatthias) [inline cloud2sql/collect_plugins.py]: I guess we can DRY this up?

**Review by aquamatthias — COMMENTED** [review]
Nice!

I suggest taking a look into #9 
Maybe it is not strictly necessary for parquet - I try to group the nodes and edges by kind, so you get a list of all entries that should be written to a table/file. 
I assume that might also be beneficial in this case?

**Review by lloesche — DISMISSED** [review]

File: cloud2sql/__main__.py [hunk 7]
```diff
@@ -34,9 +33,10 @@ def parse_args() -> Namespace:
     )
     parser.add_argument(
         "--db",
-        help="The database url. See https://docs.sqlalchemy.org/en/20/core/engines.html.",
-        required=True,
+        help="The database url or a parquet path. See https://docs.sqlalchemy.org/en/20/core/engines.html.",
+        required=False,
```
Comment (aquamatthias) [inline cloud2sql/__main__.py]: Isn't it still required?

File: cloud2sql/__main__.py [hunk 8]
```diff
@@ -34,9 +33,10 @@ def parse_args() -> Namespace:
...
     )
+    parser.add_argument("--parquet_batch_size", type=int, default=100_000, help="Batch rows before writing to parquet")
```
Comment (aquamatthias) [inline cloud2sql/__main__.py]: To make it consistent with other drivers, it should be part of the url.
```
parquet:///path/to/directory?batch-size=100000
```

File: cloud2sql/collect_plugins.py [hunk 9]
```diff
@@ -58,6 +63,64 @@ def configure(path_to_config: Optional[str]) -> Json:
...
+def collect_parquet(collector: BaseCollectorPlugin, feedback: CoreFeedback, args: Namespace) -> Tuple[str, int, int]:
+    # collect cloud data
+    feedback.progress_done(collector.cloud, 0, 1)
+    collector.collect()
+    # read the kinds created from this collector
+    kinds = [from_json(m, Kind) for m in collector.graph.export_model(walk_subclasses=False)]
+    model = ParquetModel(Model({k.fqn: k for k in kinds}))
+    node_edge_count = len(collector.graph.nodes) + len(collector.graph.edges)
+    ne_current = 0
+    progress_update = 5000
```
Comment (aquamatthias) [inline cloud2sql/collect_plugins.py]:
```suggestion
    progress_update = node_edge_count // 100
```

File: cloud2sql/collect_plugins.py [hunk 10]
```diff
@@ -58,6 +63,64 @@ def configure(path_to_config: Optional[str]) -> Json:
...
+    writer = ParquetWriter(model, parquet_path, args.parquet_batch_size)
+    node: BaseResource
+    for node in sorted(collector.graph.nodes, key=lambda n: n.kind):
+        exported = prepare_node(node, collector)
+        writer.insert_node(exported)
+        ne_current += 1
+        if ne_current % progress_update == 0:
+            feedback.progress_done("sync_db", ne_current, node_edge_count, context=[collector.cloud])
+    for from_node, to_node, _ in collector.graph.edges:
+        writer.insert_node({"from": from_node.chksum, "to": to_node.chksum, "type": "edge"})
```
Comment (aquamatthias) [inline cloud2sql/collect_plugins.py]: you need to filter out delete edges

File: cloud2sql/parquet.py [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+            # create link table for all linked entities
+            for successor in successors:
+                link_table_schema(kind.fqn, successor)
+
+        # step 1: create tables for all kinds
+        for kind in self.table_kinds:
+            table_schema(kind)
+        # step 2: create link tables for all kinds
+        for kind in self.table_kinds:
+            link_table_schema_from_successors(kind)
```
Comment (aquamatthias) [inline cloud2sql/parquet.py:81]: You also want to create link tables for the edges found in the import.
Ideally, all of the successors are defined in the model - but the reality is different...

**Review by aquamatthias — COMMENTED** [review]
Thanks! Please find my comments

File: cloud2sql/__main__.py (hunk 8 above)
Comment (meln1k) [reply cloud2sql/__main__.py]: Lukas made a change using a config mechanism. I adapted my code as well.

File: cloud2sql/collect_plugins.py [hunk 12]
```diff
@@ -65,6 +70,74 @@ def configure(path_to_config: Optional[str]) -> Json:
...
+    # read the kinds created from this collector
+    kinds = [from_json(m, Kind) for m in collector.graph.export_model(walk_subclasses=False)]
+    model = ParquetModel(Model({k.fqn: k for k in kinds}))
+    node_edge_count = len(collector.graph.nodes) + len(collector.graph.edges)
+    ne_current = 0
+    progress_update = node_edge_count // 100
+    feedback.progress_done("sync_db", 0, node_edge_count, context=[collector.cloud])
+
+    # group all edges by kind of from/to
+    edges_by_kind: Dict[Tuple[str, str], List[Json]] = defaultdict(list)
```
Comment (aquamatthias) [inline cloud2sql/collect_plugins.py]: The edge group lists are not used. Use a set to collect the kinds.

File: cloud2sql/parquet.py [hunk 13]
```diff
@@ -0,0 +1,158 @@
...
+        # step 3: create link tables for all seen edges
+        for from_kind, to_kind in edges:
+            link_table_schema(from_kind, to_kind)
+
+        return None
+
+
+class WriteResult(NamedTuple):
```
Comment (aquamatthias) [inline cloud2sql/parquet.py:89]: what's the advantage of being a named tuple?

File: cloud2sql/parquet.py [hunk 14]
```diff
@@ -0,0 +1,158 @@
...
+
+            batch.rows.append(values)
+            self.batches[table_name] = batch
+            return WriteResult(table_name)
+        return None
+
+    def write_batch_bundle(self, batch: ParquetBatch) -> None:
+        pa_table = pa.Table.from_pylist(batch.rows, batch.schema)
+        batch.writer.write_table(pa_table)
+        batch.rows = []
```
Comment (aquamatthias) [inline cloud2sql/parquet.py]: I guess there is no concurrency here? Consider assigning the rows to a local variable and emptying the batch as first action.

**Review by aquamatthias — DISMISSED** [review]
Thanks! Only some small questions/improvments.

File: cloud2sql/parquet.py (hunk 13 above)
Comment (meln1k) [reply cloud2sql/parquet.py:89]: That is clearer than returning a plain string. Otherwise, it is too easy to mix something up.

Comment (aquamatthias) [reply cloud2sql/parquet.py:89]: Not saying anything against the class - but why it inherits from NamedTuple?

File: cloud2sql/parquet.py (hunk 14 above)
Comment (meln1k) [reply cloud2sql/parquet.py]: This is single threaded, yes.

**Review by aquamatthias — APPROVED** [review]
Thanks!

File: cloud2sql/parquet.py (hunk 13 above)
Comment (meln1k) [reply cloud2sql/parquet.py:89]: just a faster and smaller thing compared to a dataclass.

**Pull request merged**